tools/codegen/gen.py

```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
         else:
             assert_never(target)
 
-    return func
+    def structured_func(g: NativeFunctionGroup) -> List[str]:
+        assert g.out is not None
+
+        # Illegal to not have dispatch for structured kernels
+        with native_function_manager(g.out):
+            assert g.out.dispatch is not None, "dispatch on structured kernels currently mandatory"
+
+            # Skip if we're generating for an unsupported dispatch key
+            if dispatch is not None and dispatch not in g.out.dispatch:
+                return []
+            # Only TypeDefault generation we do is registration
+            if dispatch is None and target is not Target.REGISTRATION:
+                return []
+
+            functional_func = g.out.func.signature()
+            functional_sig = DispatcherSignature.from_schema(functional_func)
+            meta_name = meta.name(functional_func)
+
+            # This is a little abusive; this assumes that the functionalization
+            # transformation ALWAYS refers to valid arguments in the original
+            # signature
+            functional_exprs = ', '.join(e.expr for e in functional_sig.exprs())
+
+        @with_native_function
+        def subfunc(f: NativeFunction) -> str:
+            k = f.func.kind()
+            sig = NativeSignature.from_schema(f.func)
+
+            if target is Target.DECLARATION:
+                # Not necessary!  Only TypeDefault is actually used, see
+                # https://github.com/pytorch/pytorch/issues/46319
+                return ""
```

Comment:
Maybe have subfunc return Optional[str] to fit the pattern of the other compute methods, if there's no semantic difference between empty string and None?

Reply:
Mooted as we no longer gen declarations.